[DXIL] Add DXIL version-specific TableGen specification and implementation of DXIL Ops

bharadwajy opened the pull request:
Update TableGen specification of DXIL Op records in DXIL.td per the current design document.

  - Facilitate specification of overloads, shader stage and attributes predicated on DXIL Ops predicated DXIL version.

Implement functionality to consume in TableGen backend, DXILEmitter, the above specification enhancements, and generate C++ code (in (DXILOperations.inc) that represents properties of DXIL Ops, associated type declarations and corresponding accessor functions.

Changes to DXIL Op Lowering pass to consume the DXIL Op representation generated by the TableGen back end.

Add mtriple with the required shader model version to commandline of tests.

Commit:
Update TableGen specification of DXIL Op records in DXIL.td per the current
design document.
  - Facilitate specification of overload and shader stage constraints on DXIL
    Ops predicated on Shader Model version.
  - Facilitate specification of unconstrained attributes, if needed.
  - Delete class DXILType and its concrete records as they are no longer
    needed.

Implement functionality to consume in TableGen backend, DXILEmitter, the
above specification enhancements, and generate C++ code (in (DXILOperations.inc)
that represents properties of DXIL Ops, associated type declarations and
corresponding accessor functions.

Add mtriple with the required shader model version to commandline of tests.

python3kgae commented on llvm/lib/Target/DirectX/DXILOpLowering.cpp:
```diff
@@ -72,10 +75,39 @@ static SmallVector<Value *> argVectorFlatten(CallInst *Orig,
+static VersionTuple getShaderModelVer(Module &M) {
+  std::string TTStr = M.getTargetTriple();
+  std::string Error;
+  auto Target = TargetRegistry::lookupTarget(TTStr, Error);
```
We should already be in DirectX backend now.
Is it possible to have illegal TargetTriple?

bharadwajy replied:
> We should already be in DirectX backend now. Is it possible to have illegal TargetTriple?

Deleted the redundant lookup and check in in-lined functionality.

python3kgae commented on llvm/lib/Target/DirectX/DXILOpLowering.cpp:
```diff
@@ -72,10 +75,39 @@ static SmallVector<Value *> argVectorFlatten(CallInst *Orig,
   return NewOperands;
 }
 
+static VersionTuple getShaderModelVer(Module &M) {
```
Both getShaderModelVer and getShaderKind will be used in many places, could we have them shared instead of a static function in DXILOpLowering.cpp?

bharadwajy replied:
> Both getShaderModelVer and getShaderKind will be used in many places, could we have them shared instead of a static function in DXILOpLowering.cpp?

Deleted the functions as each one is called once and in-lined the functionality.

python3kgae commented on llvm/lib/Target/DirectX/DXILOpBuilder.cpp:
```diff
@@ -249,19 +191,59 @@ static FunctionType *getDXILOpFunctionType(const OpCodeProperty *Prop,
+
+  if (ValidShaderKindMask == ShaderKind::Unknown) {
+    report_fatal_error(
+        StringRef(SMVer.getAsString().append(": Unhandled Shader Kind")),
```
Is this Unhandled Shader Kind in current SMVer?
Like use mesh shader in shader model 6.0?

bharadwajy replied:
> Is this Unhandled Shader Kind in current SMVer? Like use mesh shader in shader model 6.0?

This tests for non-zero mask value in the `OpcodeProperty` of `Opcode`.

python3kgae replied:
The error message is SMVer.getAsString().append(": Unhandled Shader Kind").
Should it be something like
ShaderKind.getAsString().append(": Unhandled Shader Kind") or
SMVer.getString().append(": Unhandled Shader Model Version") or something else?

bharadwajy replied:
> The error message is SMVer.getAsString().append(": Unhandled Shader Kind"). Should it be something like ShaderKind.getAsString().append(": Unhandled Shader Kind") or SMVer.getString().append(": Unhandled Shader Model Version") or something else?

Updated the message to emit the following (assume SM version is 6.0 and operation is `Cos`):
`LLVM ERROR: 6.0: Unknown Target Shader Stage for DXIL operation - Cos`
to signify unknown stage specification for operation `Cos` for Shader Model version 6.0

python3kgae replied:
I see.
How about something like "Cos is not supported in Shader Model 6.0" ?

bharadwajy replied:
This check ensures that shader kind mask created by `DXILEmitter` in `DXILOperations.inc` from the stage constraints specified for an operation in `DXIL.td` is not `0` (i.e., `ShaderKind::Unknown`). With the choice of wording currently proposed in my PR, my expectation is to distinguish between that and the case where an operation is not supported - which I think your suggestion would mean.

python3kgae commented on llvm/lib/Target/DirectX/DXILOpBuilder.cpp:
```diff
@@ -249,19 +191,59 @@ static FunctionType *getDXILOpFunctionType(const OpCodeProperty *Prop,
+  }
+
+  if (!(ValidShaderKindMask | getShaderkKindEnum(StageKind))) {
+    report_fatal_error(StringRef(std::string(StageKind).append(
```
Do we have tests for these errors?

bogner replied:
Agreed, we need tests for some of the error cases if they're reachable from user provided IR. I also suspect these errors are going to be pretty unhelpful since they don't say anything about what op was invalid...

bogner commented on llvm/lib/Target/DirectX/DXILOpLowering.cpp:
```diff
@@ -22,9 +22,12 @@
 #include "llvm/Pass.h"
 #include "llvm/Support/ErrorHandling.h"
 
+#include "iostream"
```
We do not use iostream in llvm: https://llvm.org/docs/CodingStandards.html#include-iostream-is-forbidden

bharadwajy replied:
> We do not use iostream in llvm: https://llvm.org/docs/CodingStandards.html#include-iostream-is-forbidden

Deleted.

bogner reviewed (commented):
A few comments about the contents of DXIL.td. I'll take a look at the tablegen logic later.

Commit:
Simplify getting shader version and shader stage kind information
from DXIL Module.

Add new tests to ensure shader stage validation checks are triggered
when DXIL Operations are created.

Clean up unused class field in DXIL.td.

Commit:
Provide additional information in error messages

bharadwajy commented on llvm/lib/Target/DirectX/DXIL.td:
```diff
@@ -212,154 +242,576 @@ defset list<DXILOpClass> OpClasses = {
+  let OpClass = isSpecialFloat;
+  let LLVMIntrinsic = int_dx_isinf;
+  let OpOverloadTypes = [llvm_half_ty, llvm_float_ty];
+  let arguments = (ins llvm_anyfloat_ty);
```
> What does it mean if the arguments don't have a `dxil_overload_ty` but we specify overloads? I guess this is a mistake.

Missed deleting the unused field, `OpOverloadTypes`. Fixed.

Commit:
Delete unused resourceVT and dxil_resource_ty

bharadwajy commented on llvm/lib/Target/DirectX/DXIL.td:
```diff
@@ -13,15 +13,45 @@
+}
+
+// Resource ValueType - has no size or value
+def resourceVT : ValueType<-1, -1>;
```
Deleted unused definitions of `resourceVT` and `dxil_resource_ty`.

bharadwajy commented on llvm/lib/Target/DirectX/DXIL.td:
```diff
@@ -212,154 +242,576 @@ defset list<DXILOpClass> OpClasses = {
+  let sm_constraints = [SMVersionConstraints<SM6_0,
+                          (overloads llvm_half_ty, llvm_float_ty),
+                          (stages allKinds)>];
+  let attributes = (attrs ReadNone);
```
> Why is this the only op with any attributes? I can't imagine `sin` being different from `cos` here...

Added attributes to other DXIL Ops.

bharadwajy commented on llvm/lib/Target/DirectX/DXIL.td:
```diff
@@ -212,154 +242,576 @@ defset list<DXILOpClass> OpClasses = {
+// Shader stages
+class ShaderStage;
+
+defset list<ShaderStage> ShaderStages = {
```
> What is this list for? It doesn't seem to be used for anything.

It is used to collect defined stage records - primarily as a syntactic "envelope".

bharadwajy replied:
> What is this list for? It doesn't seem to be used for anything.

Deleted.

bharadwajy commented on llvm/lib/Target/DirectX/DXIL.td:
```diff
@@ -212,154 +242,576 @@ defset list<DXILOpClass> OpClasses = {
+  def node : ShaderStage;
+  def raygeneration : ShaderStage;
+  def intersection : ShaderStage;
+  def allKinds : ShaderStage;
```
> Is the "allKinds" stage really necessary? Isn't that equivalent to simply not having a constraint?

Yes. I just opted to provide a way to be explicit in the description - though I am equally happy to treat non-specification of a constraint, as applying to all stages.

Commit:
Add convenience TableGen class DXILOpAndClass<>
Accomdate for non-specification of LLVMIntrinsic field

Replace use of dag with list in specifications throughout
DXIL.td

Rename getOverloadTy() to getOverloadType()

Delete all defsets from DXIL.td as they are not used
Delete datalayout attribute from newly added tests
Delete pseudo shader stage kind allKinds from DXIL.td

bharadwajy commented on llvm/lib/Target/DirectX/DXIL.td:
```diff
@@ -212,154 +242,576 @@ defset list<DXILOpClass> OpClasses = {
+class SMVersionConstraints<Version ver, dag oloads, dag stages> : SMVersion<ver> {
+  dag overload_types = oloads;
+  dag stage_kinds = stages;
+}
```
Changed constraint specification to use `list` representation instead of `dag`.

bharadwajy commented on llvm/lib/Target/DirectX/DXIL.td:
```diff
@@ -212,154 +242,576 @@ defset list<DXILOpClass> OpClasses = {
+  let sm_constraints = [SMVersionConstraints<SM6_0,
+                             (overloads llvm_half_ty, llvm_float_ty),
+                             (stages allKinds)>];
+}
```
> These definitions are very verbose, especially the ones that define constraints that seem to be redundant with their overload types like this one.
> 
> * Can some of the required fields be arguments to DXILOp? Specifically, OpCode and OpClass seem like good candidates, like `def IsInf : DXILOp<9, isSpecialFloat> { ... }`
> * Why do we need to specify constraints when we don't actually constrain anything?

Added a convenience TableGen class. Changes to constraint specification pushed that address the above.

bharadwajy reviewed (commented):
Changes that use `list` in place of `dag` pushed - per PR feedback. 

Will update the design document in accordance with these changes in a separate PR after this implementation PR is merged.

bharadwajy commented on llvm/lib/Target/DirectX/DXIL.td:
```diff
@@ -212,154 +242,576 @@ defset list<DXILOpClass> OpClasses = {
+  dag arguments = (ins);
+
+  // Results of the op. Default to 0 results.
+  dag result = (out);
```
> Why are these dags? They're only used as lists. Similarly for other uses of dag in this file

Replaced all uses of `dag` with `list` in DXIL Op specification.

python3kgae commented on llvm/lib/Target/DirectX/DXILOpBuilder.h:
```diff
@@ -30,13 +31,21 @@ class DXILOpBuilder {
   /// \return DXIL Op call constructed
-  CallInst *createDXILOpCall(dxil::OpCode OpCode, Type *ReturnTy,
+  CallInst *createDXILOpCall(dxil::OpCode OpCode, VersionTuple SMVer,
+                             StringRef StageKind, Type *ReturnTy,
```
Why StageKind is a StringRef instead of enum for shader stages?

bharadwajy replied:
> Why StageKind is a StringRef instead of enum for shader stages?

Do you think it is safer and/or more beneficial to pass the enum obtained by calling `getShaderkKindEnum(...)` at the call site `DXILOpLowering.cpp::lowerIntrinsic(...)`, instead of doing so this function?

python3kgae replied:
I saw the string is from Triple(TTStr).getEnvironmentName().str();
Could we use getEnvironment() instead of getEnvironmentName(), then we can convert the Triple::EnvironmentType to the enum we want.

bharadwajy replied:
> I saw the string is from Triple(TTStr).getEnvironmentName().str(); Could we use getEnvironment() instead of getEnvironmentName(), then we can convert the Triple::EnvironmentType to the enum we want.

I see what you mean. Actually, I think Shader Model version and Shader Kind (Environment enum) do not need to be passed to `DXILOpBuilder::createDXILOpCall(...)` since the function has access to the private `Module` instance. That was part of an NFC PR I was planning after this one - in which I want to move `lowerIntrinsic(...)` and a couple of other functions in `DXILOpLowering.cpp` as member functions of `class DXILOpBuilder`. I refrained from including those changes with an expectation to reduce the size/scope of this. I can extract just the change that avoids passing the values of SM version and stage kind.

bharadwajy replied:
> ... I can extract just the change that avoids passing the values of SM version and stage kind.

Pushed the change.

Commit:
Use Module field of DXILOpBuilder to find target SM version and Shader
Kind in createDXILOpCall() instead of computing them in its callers and
passing them as arguments.

python3kgae commented on llvm/lib/Target/DirectX/DXILOpBuilder.cpp:
```diff
@@ -213,11 +251,23 @@ static int getValidConstraintIndex(const OpCodeProperty *Prop,
+  // well-formed or the target is supported since these checks would have
+  // been done at the time the module M is constructed in the earlier stages of
+  // compilation.
+  auto Major = Triple(TTStr).getOSVersion().getMajor();
```
This will build Triple from TTStr 3 times every time createDXILOpCall is invoked.
Maybe we can add Triple as a member of DXILOpBuilder?

bharadwajy replied:
> This will build Triple from TTStr 3 times every time createDXILOpCall is invoked. Maybe we can add Triple as a member of DXILOpBuilder?

I expect that change to be part of the follow up NFC PR with remaining changes planned in `DXILBuilder` class that I mentioned [here](https://github.com/llvm/llvm-project/pull/97593#discussion_r1673006190).

python3kgae replied:
Got it.

python3kgae commented on llvm/lib/Target/DirectX/DXILOpBuilder.cpp:
```diff
@@ -249,17 +228,102 @@ static FunctionType *getDXILOpFunctionType(const OpCodeProperty *Prop,
+  // Ensure Environment type is known
+  if (ShaderEnv == Triple::UnknownEnvironment) {
+    report_fatal_error(
+        StringRef(SMVer.getAsString().append(
```
This error could simply check the target triple once instead of checking it every time when building DXIL operations.

bharadwajy replied:
> This error could simply check the target triple once instead of checking it every time when building DXIL operations.

Yes. Per my current design of `DXILOpBuilder` class update (in the follow-on NFC PR), that check would be done as part of its constructor where the frequently accessed / computed `Module` information - such as `TargetTriple`, for now - is initialized (potentially as private member fields with corresponding public accessor functions).

Commit:
Delete unused code snippet

damyanp commented on llvm/lib/Target/DirectX/DXIL.td:
```diff
@@ -13,353 +13,777 @@
+}
+
+// Following class represents predicate that evaluates to the payload when
+// specified predicte is true.
```
```suggestion
// specified predicate is true.
```

damyanp commented on llvm/lib/Target/DirectX/DXIL.td:
```diff
@@ -13,353 +13,777 @@
+// Following classes represent predicate that evaluate to the payload,
+// if one xists, unconditionally.
+
+// DXIL Op attributes defined as true predicates, if specified
+
+def ReadOnly : Pred;
+def ReadNone : Pred;
+def IsDerivative : Pred;
+def IsGradient : Pred;
+def IsFeedback : Pred;
+def IsWave : Pred;
+def NeedsUniformInputs : Pred;
+def IsBarrier : Pred;
```
I'm probably missing something, but how come these attributes are predicates?  Can they be used in a constraint?

bharadwajy replied:
> I'm probably missing something, but how come these attributes are predicates? Can they be used in a constraint?

Yes. It would express constraints when the attribute is true. Not yet seen a need for such a specification, though. The idea is to use a unified TableGen class to model an attribute as a predicate that is always true instead of using a separate `class Attribute` to convey the same semantics.

damyanp replied:
I'm afraid I don't follow what "constraints when the attribute is true" means.  If these attributes are always true, then what would a constraint involving one of these attributes mean?

bharadwajy replied:
As an example, consider the hypothetical example where an attribute represented as a predicate can be used in constraint (per the original question) - `Constraint<IsWave, [SomePredicate, SomeOtherPredicate]>`. It represents that if `IsWave` is true, the predicates `SomePredicate` and `SomeOtherPredicate` are true for the DXIL Operation. This is similar to the specification `Constraints<SMVersion<SM6_0>, [Overloads<[llvm_half_ty, llvm_float_ty, llvm_double_ty]>]>]` that represents that if the predicate `SMVersion<SM6_0>` is true, the predicate `Overloads<llvm_half_ty, llvm_float_ty, llvm_double_ty]>]>]` is true.

damyanp replied:
So my confusion here is that it seems that `IsWave` is defined to be always true (line 251 above).  How could it be not true?

bharadwajy replied:
A predicate - unconditional or conditional - when not specified, represents that it is not true for that operation. For example, if `IsWave` (unconditional) is specified, the operation has the property of being a wave operation. Absence of it indicates that it is a not constrained to be a wave operation. Similarly, `Constraints<SMVersion<SM6_0>, [P1, P2]>` (conditional predicate) indicates that the operation has the properties `P1` and `P2` for the predicate `SMVersion<SM<6_0>` (i.e., `SMVersion<SM6_0>` is true, by virtue of it being specified). Absence of any `Constraints<...>` indicates that the operation has no conditional predicates (i.e., properties) and is valid in all circumstances.

damyanp replied:
I think some confusion here is that this PR is now diverging quite a from the design described in DXILOpTableGenDesign - where there was a DXILAttributes class.

If we're going to be implementing a new design, maybe we need to go back and get that spec updated first so we can be sure we're all on the same page.

bharadwajy replied:
> I think some confusion here is that this PR is now diverging quite a from the design described in DXILOpTableGenDesign - where there was a DXILAttributes class.
> 
> If we're going to be implementing a new design, maybe we need to go back and get that spec updated first so we can be sure we're all on the same page.

Pushed changes in accordance with the latest design doc.

llvm-beanz commented on llvm/lib/Target/DirectX/DXIL.td:
```diff
@@ -13,353 +13,777 @@
-                                             "Provides a flattened index for a "
-                                             "given thread within a given "
-                                             "group (SV_GroupIndex)">;
+// Concrete definitions of DXIL Op Classes
```
There is a huge diff here changing how these op classes are defined, but it looks like that isn't needed now. Can we revert that to reduce the diff noise?

bharadwajy replied:
> There is a huge diff here changing how these op classes are defined, but it looks like that isn't needed now. Can we revert that to reduce the diff noise?

Definition of op classes has not changed. Prior to this change, all `DXILOpClasses` are defined as records of a list - `OpClasses`. Although the list `OpClasses` is not used, the intent was to provide a syntactic / visual "envelope". However, per feedback [here](https://github.com/llvm/llvm-project/pull/97593/files/b98b2f17ac2b92b5743ca67e5c05e039d83d52f0#r1664610418), the unused list definition is removed while retaining all the op class definitions unchanged. The difference results from this change.

damyanp commented on llvm/lib/Target/DirectX/DXILOpBuilder.cpp:
```diff
@@ -221,6 +161,45 @@ static Type *getTypeFromParameterKind(ParameterKind Kind, Type *OverloadTy) {
+  default:
+    break;
+  }
+  llvm_unreachable(
```
Should this be a diagnostic?  I notice in the place that this function is called, UnknownEnvironment will cause a diagnostic.

llvm-beanz commented on llvm/utils/TableGen/DXILEmitter.cpp:
```diff
@@ -469,28 +561,138 @@ static void emitDXILOperationTable(std::vector<DXILOperationDesc> &Ops,
+      break; \n \
+    } \n \
+    llvm_unreachable(\"invalid overload type for name\"); \n \
+  }\n\n";
```
I think it was better having this declared in source rather than emitted here. I also wonder if `OpCodeProperty` should be here or in source.

damyanp commented on llvm/lib/Target/DirectX/DXIL.td:
```diff
@@ -13,353 +13,777 @@
+// specified predicte is true.
+
+class Constraints<Pred p, list<Pred> l = []> : Pred {
+  Pred pred = p;
```
It looks likke the tablegen code assumes that `pred` is always a `SMVersion`.  Could this be `SMVersion` instead?

bharadwajy replied:
> It looks like the tablegen code assumes that `pred` is always a `SMVersion`. Could this be `SMVersion` instead?

Currently, the constraints are only based on Shader Model version. This class definition is intended to be generic enough to represent other constraints, if needed.

llvm-beanz commented on llvm/utils/TableGen/DXILEmitter.cpp:
```diff
@@ -124,32 +117,29 @@ DXILOperationDesc::DXILOperationDesc(const Record *R) {
-    auto isAny = TR->getValueAsInt("isAny");
-    if (isAny == 1) {
+    auto IsAny = TR->getValueAsInt("isAny");
+    if (IsAny == 1) {
```
There are so many NFC changes that clean up this code (which is great) but grouping those in an already large PR makes it a lot harder to review.

damyanp commented on llvm/lib/Target/DirectX/DXILOpBuilder.cpp:
```diff
@@ -249,17 +228,102 @@ static FunctionType *getDXILOpFunctionType(const OpCodeProperty *Prop,
+
+  // Ensure valid shader stage constraints are specified
+  if (ValidShaderKindMask == ShaderKind::Unknown) {
+    report_fatal_error(
```
The clang code seems to have a diagnostic td file that these messages go into.  Is it normal practice not to do that in llvm? Or is this more about the types of errors that mean they don't go into a centralized table?

llvm-beanz replied:
There isn't really an equivalent to Clang's diagnostic tablegen in LLVM. Generally LLVM's errors are all fatal so they report this way (basically exiting the compiler).

We may want to consider a larger design proposal for LLVM to allow passes to propagate errors up. This has been discussed in the past, but never really worked on.

bogner replied:
There are so-called "Backend Diagnostics" in `llvm/IR/DiagnosticInfo.h`, which is probably what we should actually be using here since it can show source locations, though it does mostly have the same hand-rolled-strings issue as this. However:
1. We've discussed making sure we have good diagnostics from the frontend for misusing APIs, so we should only get here if we're starting from LLVM IR 
2. That kind of diagnostic should really be coming out of the DXILOpLowering pass and not from inside the DXILOpBuidler anyway

For now I think the assert/fatal error approach is probably fine while we get things propped up a little better, but we will want to improve this over time.

damyanp commented on llvm/utils/TableGen/DXILEmitter.cpp:
```diff
@@ -113,9 +99,16 @@ DXILOperationDesc::DXILOperationDesc(const Record *R) {
+  for (auto Ty : ArgTys) {
+    ParamTypeRecs.push_back(Ty);
+  }
+  unsigned ParamTypeRecsSize = ParamTypeRecs.size();
```
Should this be `size_t` rather than `unsigned`?

damyanp commented on llvm/utils/TableGen/DXILEmitter.cpp:
```diff
@@ -124,32 +117,29 @@ DXILOperationDesc::DXILOperationDesc(const Record *R) {
-    if (isAny == 1) {
+    auto IsAny = TR->getValueAsInt("isAny");
+    if (IsAny == 1) {
       // TODO: At present it is expected that all overload types in a DXIL Op
       // are of the same type. Hence, OverloadParamIndices will have only one
       // element. This implies we do not need a vector. However, until more
       // (all?) DXIL Ops are added in DXIL.td, a vector is being used to flag
       // cases this assumption would not hold.
```
Is this comment out of date now?

damyanp commented on llvm/utils/TableGen/DXILEmitter.cpp:
```diff
@@ -469,28 +561,138 @@ static void emitDXILOperationTable(std::vector<DXILOperationDesc> &Ops,
+          };\n\n";
+
+  // Emit access function getOverloadTypeName()
+  OS << "static const char *getOverloadTypeName(OverloadKind Kind) { \n \
```
Any reason not to use raw string literals here so there's less escaping?

damyanp reviewed (changes requested):
We should get the spec updated to match the design being implemented here.

Commit:
Revert NFC changes to reduce changeset size.

Commit:
Declare and use TableGen class DXILAttribute instead of class Pred
to define DXIL attributes

Commit:
Bug fix

bharadwajy commented:
> We should get the spec updated to match the design being implemented here.

Spec updated - PR #99055.

Commit:
Changes in accordance with the latest update to design document
for TableGen specification of DXIL Operations.

Commit:
Use Twine to construct message in report_fatal_error()

llvm-beanz commented on llvm/lib/Target/DirectX/DXILOpBuilder.cpp:
```diff
@@ -131,8 +151,9 @@ struct OpCodeProperty {
-  llvm::Attribute::AttrKind FuncAttr;
+  std::vector<OpOverload> Overloads;
+  std::vector<OpStage> Stages;
+  std::vector<OpAttribute> Attributes;
```
We should probably use `llvm::SmallVector` for these to avoid unnecessary allocations.

bharadwajy replied:
> We should probably use `llvm::SmallVector` for these to avoid unnecessary allocations.

Changed.

llvm-beanz commented on llvm/lib/Target/DirectX/DXILOpBuilder.cpp:
```diff
@@ -249,17 +309,95 @@ static FunctionType *getDXILOpFunctionType(const OpCodeProperty *Prop,
+/// DXIL version not greater than DXILVer.
+/// PropList is expected to be sorted in ascending order of DXIL version.
+template <typename T>
+static int getPropIndex(const std::vector<T> PropList,
```
`const &` or maybe `llvm::ArrayRef<T>` instead to avoid copies.

bharadwajy replied:
> `const &` or maybe `llvm::ArrayRef<T>` instead to avoid copies.

Changed to use `llvm::ArrayRef`.

llvm-beanz commented on llvm/lib/Target/DirectX/DXILOpBuilder.cpp:
```diff
@@ -249,17 +309,95 @@ static FunctionType *getDXILOpFunctionType(const OpCodeProperty *Prop,
+template <typename T>
+static int getPropIndex(const std::vector<T> PropList,
+                        const VersionTuple DXILVer) {
+  auto Size = PropList.size();
```
`auto` doesn't help readability here (https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable)

```suggestion
  size_t Size = PropList.size();
```

bharadwajy replied:
> `auto` doesn't help readability here (https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable)

Changed `auto` to `size_t`.

llvm-beanz commented on llvm/lib/Target/DirectX/DXILOpBuilder.cpp:
```diff
@@ -249,17 +309,95 @@ static FunctionType *getDXILOpFunctionType(const OpCodeProperty *Prop,
+                        const VersionTuple DXILVer) {
+  auto Size = PropList.size();
+  for (int I = Size - 1; I >= 0; I--) {
+    auto OL = PropList[I];
```
see: https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable
```suggestion
    const T& OL = PropList[I];
```

bharadwajy replied:
> see: https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable

Changed `auto` to `const T&`

llvm-beanz commented on llvm/lib/Target/DirectX/DXILOpBuilder.cpp:
```diff
@@ -249,17 +309,95 @@ static FunctionType *getDXILOpFunctionType(const OpCodeProperty *Prop,
+      return I;
+    }
+  }
+  report_fatal_error(Twine(DXILVer.getAsString()) + ": Unknown DXIL Version",
```
Is this really the right place to throw this error? It seems like a getter is an odd place to report an error as opposed to the caller.

Also since `report_fatal_error` is `[[noreturn]]` there shouldn't be any code after it.

bharadwajy replied:
> Is this really the right place to throw this error? It seems like a getter is an odd place to report an error as opposed to the caller.
> 
> Also since `report_fatal_error` is `[[noreturn]]` there shouldn't be any code after it.

Moved error out of getter to callsite.

llvm-beanz commented on llvm/lib/Target/DirectX/DXILOpBuilder.cpp:
```diff
@@ -249,17 +309,95 @@ static FunctionType *getDXILOpFunctionType(const OpCodeProperty *Prop,
                                           Type *OverloadTy,
                                           SmallVector<Value *> Args) {
+
+  auto Major = Triple(TargetTripleStr).getDXILVersion().getMajor();
+  auto MinorOrErr = Triple(TargetTripleStr).getDXILVersion().getMinor();
+  uint32_t Minor = MinorOrErr.has_value() ? *MinorOrErr : 0;
+  VersionTuple DXILVer(Major, Minor);
```
```suggestion
  VersionTuple DXILVer = Triple(TargetTripleStr).getDXILVersion();
```

bharadwajy replied:
Done.

llvm-beanz commented on llvm/lib/Target/DirectX/DXILOpBuilder.cpp:
```diff
@@ -249,17 +309,95 @@ static FunctionType *getDXILOpFunctionType(const OpCodeProperty *Prop,
+  // kind
+  int StIndex = getPropIndex(Prop->Stages, DXILVer);
+  uint16_t ValidShaderKindMask = Prop->Stages[StIndex].ValidStages;
+  enum ShaderKind ModuleStagekind = getShaderKindEnum(ShaderEnv);
```
```suggestion
  ShaderKind ModuleStagekind = getShaderKindEnum(ShaderEnv);
```

bharadwajy replied:
Changed `enum ShaderKind` declaration to `ShaderKind`.

llvm-beanz commented on llvm/lib/Target/DirectX/DXILOpBuilder.cpp:
```diff
@@ -249,17 +309,95 @@ static FunctionType *getDXILOpFunctionType(const OpCodeProperty *Prop,
+
+  // Verify the target shader stage is valid for the DXIL operation
+  if (!(ValidShaderKindMask & ModuleStagekind)) {
+    auto ShaderEnvStr = Triple(TargetTripleStr).getEnvironmentName();
```
You have a bunch of places that do this `Triple(TargetTripleStr)` over and over again. Should you just cache the constructed triple?
```suggestion
    StringRef ShaderEnvStr = Triple(TargetTripleStr).getEnvironmentName();
```
```suggestion
    auto ShaderEnvStr = Triple(TargetTripleStr).getEnvironmentName();
```

bharadwajy replied:
> You have a bunch of places that do this `Triple(TargetTripleStr)` over and over again. Should you just cache the constructed triple?

Replaced `std::string DXILOpBuilder::TargetTripleStr` with `Triple::TargetTriple` to cache the Triple.

llvm-beanz commented on llvm/lib/Target/DirectX/DXILOpBuilder.cpp:
```diff
@@ -282,40 +420,18 @@ Type *DXILOpBuilder::getOverloadTy(dxil::OpCode OpCode, FunctionType *FT) {
   }
 
-  auto ParamKinds = getOpCodeParameterKind(*Prop);
+  const auto *ParamKinds = getOpCodeParameterKind(*Prop);
```
I don't know what this type is from the context here which makes me think it shouldn't be `auto` (https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable).

bharadwajy replied:
> I don't know what this type is from the context here which makes me think it shouldn't be `auto` (https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable).

Changed `auto` to `ParamKind`

Commit:
Changes that address PR Feedback

bharadwajy reviewed (commented):
Changes  pushed to address all PR feedback.

bogner commented on llvm/lib/Target/DirectX/DXILOpBuilder.h:
```diff
@@ -42,6 +48,7 @@ class DXILOpBuilder {
 private:
   Module &M;
   IRBuilderBase &B;
+  Triple TargetTriple;
```
It will simplify things to have the DXIL version and shader stage as members instead of the triple, and then we can just assume that they're valid like so:
```c++
DXILOpBuilder::DXILOpBuilder(Module &M, IRBuilderBase &B) : M(M), B(B) {
  Triple TT(M.getTargetTriple());
  DXILVersion = TT.getDXILVersion();
  assert(TT.isShaderStageEnvironment() &&
         "Cannot build DXIL ops outside of shader environment");
  ShaderStage = TT.getEnvironment();
}
```
It shouldn't be DXILOpBuilder's job to validate the triple.

bharadwajy replied:
> It will simplify things to have the DXIL version and shader stage as members instead of the triple, and then we can just assume that they're valid like so:
> 
> ```c++
> DXILOpBuilder::DXILOpBuilder(Module &M, IRBuilderBase &B) : M(M), B(B) {
>   Triple TT(M.getTargetTriple());
>   DXILVersion = TT.getDXILVersion();
>   assert(TT.isShaderStageEnvironment() &&
>          "Cannot build DXIL ops outside of shader environment");
>   ShaderStage = TT.getEnvironment();
> }
> ```
> 
> It shouldn't be DXILOpBuilder's job to validate the triple.

Changed.

bogner reviewed (approved):
This looks mostly okay. The error handling is awkward and the way that `OpCodeProperty` is defined and used is not very idiomatic tablegen and the whole thing is fairly unwieldy, which are both things we'll need to fix. This change is just adding to the existing pattern though, so it's probably best to deal with that in follow ups.

Commit:
Changes from PR feedback

- Delete anyFloatTy; Replace report_fatal_error with PrintFatalError

- Stash DXILVersion and ShaderStage information from Module in DXILOpBuilder object.

bharadwajy commented on llvm/lib/Target/DirectX/DXIL.td:
```diff
@@ -13,14 +13,33 @@
+defvar halfTy = llvm_half_ty;
+defvar floatTy = llvm_float_ty;
+defvar doubleTy = llvm_double_ty;
+defvar anyfloatTy = llvm_anyfloat_ty;
```
> I don't think we need `anyfloatTy` - everywhere that it's used currently is really just `overloadTy`

Replaced `anyfloatTy` with `overloadTy`.

bharadwajy commented on llvm/utils/TableGen/DXILEmitter.cpp:
```diff
@@ -170,28 +171,64 @@ DXILOperationDesc::DXILOperationDesc(const Record *R) {
+  if (Recs.empty()) {
+    report_fatal_error(Twine("Atleast one specification of valid stage for ") +
+                           OpName + " is required",
+                       /* gen_crash_diag=*/false);
```
> "llvm/TableGen/Error.h" has `PrintFatalError`, which is better because it can print a source location. We should use that for all of the errors in DXILEmitter.

Changed usage of `report_fatal_error()` to `PrintFatalError()`
